Decoder and list combinator

cryogenian opened the pull request:
* Added functions to provide a way to explicit set decoder for uri (not only `decodeURIComponent`)
* Added `list` combinator that matches provided matcher as argument many times and yields `Match (List a)`

jdegoes commented on src/Routing.purs:
```diff
@@ -29,11 +39,18 @@ hashes cb =
 
 
 matches :: forall e a. Match a -> (Maybe a -> a -> Eff e Unit) -> Eff e Unit
```
This `Eff` feels like it doesn't belong here. You have a few options:

```purescript
matches :: forall e a b. Match a -> (Maybe a -> a -> b) -> Maybe b

matches :: forall e a b. (Monoid b) => Match a -> (Maybe a -> a -> b) -> b
```

With the 2nd option, you'd have to submit a PR for `purescript-monoid` to define `Monoid (Eff eff a)` when `a` is a `Monoid` (which is a sensible monoid to have, just missing).

jdegoes commented on src/Routing/Match.purs:
```diff
@@ -21,6 +21,8 @@ import Routing.Types
 import Routing.Match.Class
 import Routing.Match.Error
 
+import Debug.Foreign
```
Is this import actually used?

cryogenian commented:
* removed `Debug.Foreign`
* added comment about `matches` 

Don't sure that I understand why there is shouldn't be two `Eff` in matches. Second argument is effectful callback, that called after hash changed and can be matched with first argument. If previous hash can be matched too, then first argument of callback is `Just a` else -- `Nothing`

jdegoes commented:
The point is that the type signature is *extremely specific* and can be made more generic. Generic code is *easier to get right* (there are fewer ways it can go wrong) and is useful solving *strictly more* problems. The only reason `matches` / `matches'` wants to be specific is so it can create an "empty", so there are really two possibilities to make it more generic: require an empty (`Monoid`) or push that off on the user (return a `Maybe`).

```purescript
matches1 :: forall e a b. (Monoid b) => Match a -> (Maybe a -> a -> b) -> b

matches2 :: forall e a b. Match a -> (Maybe a -> a -> b) -> Maybe b
```

Look at how the old one can be implemented in terms of either of these:

```purescript
matches3 m f = matches1 m f -- assuming monoid for Eff eff Unit

matches4 m f = fromMaybe (pure unit) (matches2 m f )
```

In addition to these points, pure code shouldn't require effects when it doesn't have to. It's very surprising to see a matching function expect an effectful callback, it implies the matcher function is going to do something effectful, too, which is not the case.
